tests/manager_arg_passer.c

```diff
@@ -0,0 +1,51 @@
+#include "manager.h"
+
+/* Test wrapper for compartments with arguments.
+ * TODO
+ *
+ * Takes at least two arguments: one argument for the compartment binary name,
+ * and one argument minimum to be passed to the compartment itself. This test
+ * only targets the `main` function of compartments.
+ */
+
+extern struct Compartment* loaded_comp;
+
+int
+main(int argc, char** argv)
+{
+    // Initial setup
+    manager_ddc = cheri_ddc_get();
+    setup_intercepts();
+
+    assert(argc >= 3 && "Expect at least two arguments: binary file for compartment, and at least one argument to pass to compartment.");
+    char* file = argv[1];
+
+    // TODO
+    char* arg_comp_entry_point[1] = { "check_fn" };
+    const size_t arg_count = 2;
+    int* arg_one = malloc(sizeof(int));
+    *arg_one = 40;
+    int* arg_two = malloc(sizeof(int));
+    *arg_two = 2;
+    void** arg_comp_args = malloc(arg_count * sizeof(void*));
```

Comment:
This might be a dumb question, but why we are we doing `malloc(malloc())` for an int? Could we not do `arg_comp_args = malloc(sizeof(int) * 2)` and avoid the double boxing?

Reply:
Right, so here's one point of contention. I would like to pass a single argument to `comp_exec` [1], which is a type-agnostic blob of arguments. The only way I could think of doing this is having `void` pointers to the data I will be packing, which will "just be read right" by the compartment function (or further validated). So I need an array of pointers to arguments.

Now, if I were to just declare my variables on my stack and take those addresses, I don't think there's any guarantee those addresses are valid when I get in `comp_exec` (at least that's what I think happened in some of my crashes, basically had a pointer to now invalid memory when entering the call, but did not properly trace where the pointer was invalidated). So I need to get heap addresses to ensure their continuity, explaining the double `malloc`.

[1] https://github.com/capablevms/CHERI-ELF-comp/pull/6/files#diff-6e0cbcab5be8419f8d8accdd49cf5e8096fefb3f0f6f3105b8cb95662ba16347R43

Reply:
Passing a single blob seems fine, but does that blob have to be a blob containing pointers? Can it just be a `int *` blob?

Reply:
I think that works for the one test right now with two `int` parameters, but don't think that works if I want to pass in like an `int`, a `float`, and a `char`. I'll add some more tests right now and play around.

Reply:
You can always treat it as `int_u8 *` i.e. just a block of bytes and pad it (or use a struct) as appropriate. That said, at some point an array of pointers is easier.

Reply:
I've added two more tests in https://github.com/capablevms/CHERI-ELF-comp/pull/6/commits/09e80b02193bd0b9d24a72a85c4fe73005406db2. I think using a struct on the manager end would entail knowing what the compartment expects (which might not be unreasonable for some use cases). I've also played around with unions. I think a union of all expected types to be sent, then have an array of those would be possible (but potentially waste a lot of padding space?)

Now that I remembered that `memcpy` exists however, I can play around with that option, with a single `malloc`. Should I try that a bit?

Reply:
I think this is probably enough for one PR?

Reply:
Alright; I'll leave playing with a single `malloc` for a later PR.